### PR TITLE
net/nettest: make nettest.NewConn pass x/net/nettest.TestConn

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Review by sailorfrag — APPROVED** [review]
This looks reasonable to me, but I am certainly not an expert on sync.Cond and haven't used this package before, so take my opinion with a grain of salt.

File: net/nettest/conn.go [hunk 1]
```diff
@@ -5,21 +5,13 @@
 package nettest
 
 import (
-	"io"
+	"net"
 	"time"
 )
 
 // Conn is a bi-directional in-memory stream that looks like a TCP net.Conn.
 type Conn interface {
```
Comment (bradfitz) [inline net/nettest/conn.go:13]: aside: this is a very useful type and not entirely related to testing (at least how we plan to use it), but I guess it works here for now.

File: net/nettest/conn.go [hunk 2]
```diff
@@ -5,21 +5,13 @@
 package nettest
 
 import (
-	"io"
+	"net"
 	"time"
 )
 
 // Conn is a bi-directional in-memory stream that looks like a TCP net.Conn.
```
Comment (bradfitz) [inline net/nettest/conn.go:12]: this isn't really how you document an interface, but this was here before.

How about:

// Conn is a net.Conn that can additionally have its reads and writes blocked and unblocked.

File: net/nettest/conn.go [hunk 3]
```diff
@@ -40,24 +32,40 @@ func NewConn(name string, maxBuf int) (Conn, Conn) {
 	return &connHalf{r: r, w: w}, &connHalf{r: w, w: r}
 }
 
+type connAddr struct {
```
Comment (bradfitz) [inline net/nettest/conn.go:35]: could just be `type connAddr string`?

File: net/nettest/pipe.go [hunk 4]
```diff
@@ -96,47 +88,58 @@ func (p *Pipe) Write(b []byte) (n int, err error) {
...
+		p.buf.Write(b[:n2])
+		b = b[n2:]
+		n += n2
+		p.cnd.L.Unlock()
+		p.cnd.Signal()
 	}
+	return n, nil
+}
+
+func (p *Pipe) EOF() {
```
Comment (bradfitz) [inline net/nettest/pipe.go:129]: this is neither documented nor obvious

Comment (bradfitz) [reply net/nettest/pipe.go:129]: it also sounds like an accessor method and not something with side effects.

File: net/nettest/pipe.go [hunk 5]
```diff
@@ -38,17 +39,10 @@ func NewPipe(name string, maxBuf int) *Pipe {
 	return &Pipe{
 		name:   name,
 		maxBuf: maxBuf,
-		rCh:    make(chan struct{}, 1),
-		wCh:    make(chan struct{}, 1),
+		cnd:    sync.NewCond(&sync.Mutex{}),
```
Comment (bradfitz) [inline net/nettest/pipe.go:42]: I wouldn't allocate a new mutex. I'd just add a `mu sync.Mutex` field to Pipe and here do:

```
p := &Pipe{..}
p.cnd = sync.NewCond(&p.mu)
```

then the `p.mu.Lock() + defer p.mu.Unlock()` look more normal

File: net/nettest/pipe.go [hunk 6]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
+		case <-ctx.Done():
+		}
+	}()
+	return cancel
+}
+
 // SetReadDeadline sets the deadline for future Read calls.
 func (p *Pipe) SetReadDeadline(t time.Time) error {
-	p.mu.Lock()
+	p.cnd.L.Lock()
```
Comment (bradfitz) [inline net/nettest/pipe.go:185]: Prefer defer Unlock whenever possible

p.mu.Lock + defer p.mu.Unlock

File: net/nettest/pipe.go [hunk 7]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
-	p.signalRead()
+	p.cancelReadTimer = p.deadlineTimer(t)
+	p.cnd.L.Unlock()
 	return nil
 }
 
 // SetWriteDeadline sets the deadline for future Write calls.
 func (p *Pipe) SetWriteDeadline(t time.Time) error {
-	p.mu.Lock()
+	p.cnd.L.Lock()
```
Comment (bradfitz) [inline net/nettest/pipe.go:198]: same

File: net/nettest/pipe.go [hunk 8]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
-
-	p.signalWrite()
+	p.cancelWriteTimer = p.deadlineTimer(t)
+	p.cnd.L.Unlock()
 	return nil
 }
 
 func (p *Pipe) Block() error {
-	p.mu.Lock()
+	p.cnd.L.Lock()
```
Comment (bradfitz) [inline net/nettest/pipe.go:210]: even here; it's okay to Broadcast with the lock held:

> It is allowed but not required for the caller to hold c.L during the call.

(from docs)

File: net/nettest/pipe.go [hunk 9]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
-	}
-	p.mu.Unlock()
-
-	p.signalWrite()
+	p.cancelWriteTimer = p.deadlineTimer(t)
+	p.cnd.L.Unlock()
 	return nil
 }
 
 func (p *Pipe) Block() error {
```
Comment (bradfitz) [inline net/nettest/pipe.go:209]: add docs while you're here?

File: net/nettest/pipe.go [hunk 10]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
+func (p *Pipe) deadlineTimer(t time.Time) func() {
+	if t.IsZero() {
+		return nil
+	}
+	d := time.Until(t)
+	if d < 0 {
+		p.cnd.Broadcast()
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (bradfitz) [inline net/nettest/pipe.go:170]: use https://golang.org/pkg/context/#WithDeadline instead and ditch time.Until and time.NewTimer?

File: net/nettest/pipe.go [hunk 11]
```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
...
+	p.cnd.Broadcast()
 
-	p.signalRead()
-	p.signalWrite()
 	return nil
 }
 
+func (p *Pipe) deadlineTimer(t time.Time) func() {
+	if t.IsZero() {
+		return nil
```
Comment (bradfitz) [inline net/nettest/pipe.go:163]: ideally you'd also track the outstanding goroutine from before and shut it down.

likewise when the pipe is closed.

Comment (bradfitz) [reply net/nettest/pipe.go:163]: oh! you do!

I see now.

File: net/nettest/conn.go [hunk 12]
```diff
@@ -40,24 +32,38 @@ func NewConn(name string, maxBuf int) (Conn, Conn) {
...
 func (c *connHalf) Close() error {
-	err1 := c.r.Close()
-	err2 := c.w.Close()
-	if err1 != nil {
-		return err1
+	if err := c.w.Close(); err != nil {
+		return err
 	}
-	return err2
+	c.r.CloseWrite()
```
Comment (bradfitz) [inline net/nettest/conn.go:63]: what's with this part of the change? Why did this Close change to a real Close in one direction but a CloseWrite in the other?

File: net/nettest/listener.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	"context"
+	"net"
+	"sync"
+)
+
+const (
+	bufferSize = 256 * 1024
+)
+
+// Listener is a net.Listener using Conn.
```
Comment (bradfitz) [inline net/nettest/listener.go:17]: but Conn's just an interface. I'd say "... using NewConn to create pairs of network connections connected in memory using a buffered pipe".

File: net/nettest/listener.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, net.ErrClosed
+	}
+}
+
+// Dial connects to the listener using the provided context.
+// The provided Context must be non-nil. If the context expires before the
+// connection is complete, an error is returned. Once successfully connected
+// any expiration of the context will not affect the connection.
+func (l *Listener) Dial(ctx context.Context, network, addr string) (net.Conn, error) {
+	if connAddr(addr) != l.addr {
```
Comment (bradfitz) [inline net/nettest/listener.go:63]: maybe also check that `strings.HasSuffix(network, "tcp")` ?

**Review by bradfitz — APPROVED** [review]
This is looking good!

File: net/nettest/conn.go (hunk 12 above)
Comment (maisem) [reply net/nettest/conn.go:63]: IIUC in a client/server interaction, connections are usually closed from a single side. The other side doesn't see anything until the read buffer is drained. With the initial implementation we would see error when the other side closed the connection.
However I can simplify this just by returning nil from Close if the pipe has already been closed.

**Comment by bradfitz** [comment]
LGTM

**Pull request merged**